examples/cvector-generator/vanilla_pca.hpp

```diff
@@ -0,0 +1,314 @@
+    if (!with_data) return;
+    printf("%s: %s[0] = [", __func__, t->name);
+    for (size_t i = 0; i <= DEBUG_POS; i++) {
+        printf(" %f,", ggml_get_f32_nd(t, i, 0, 0, 0));
+    }
+    printf(" ... ]\n");
+}
+
+// begin vanilla pca namespace
+namespace PCA {
+
+// input params for PCA computations
+struct pca_params {
+    int n_threads    = 1;
+    int n_batch      = 20; // number of iterations do to in one batch. larger the batch, more memory is used
+    int n_iterations = 1000;
+    float tolerance  = 1e-7;
+};
+
+// result from each iteration
+struct pca_result {
+    struct ggml_tensor * principal_component; // eigenvectors of the covariance matrix
+    float explained_variance;                 // eigenvalues of the covariance matrix
+};
+
+void compute_covariance(struct pca_params &pca_params,
+                        struct ggml_tensor * X,
+                        struct ggml_tensor * covariance,
+                        struct ggml_backend * backend) {
+
+    // Memory allocation
+    struct ggml_cgraph  * gf  = NULL;
+    struct ggml_context * ctx = NULL;
+    struct ggml_init_params ctx_params = {
+        ggml_tensor_overhead()*GGML_DEFAULT_GRAPH_SIZE + ggml_graph_overhead(),
+        NULL,
+        true, // the tensors will be allocated later by ggml_gallocr_alloc_graph()
+    };
+    ctx = ggml_init(ctx_params);
+    gf  = ggml_new_graph(ctx);
+
+    // Step 0: Transpose the input because of row-major
+    X = ggml_cont(ctx, ggml_transpose(ctx, X));
+
+    // Step 1: Compute the mean for each feature
+    struct ggml_tensor * mean           = ggml_repeat(ctx, ggml_mean(ctx, X), X); // mean with trick to make it easier to sub
+    struct ggml_tensor * centered_data  = ggml_sub(ctx, X, mean);
+
+    // Step 2: Compute the covariance matrix
+    struct ggml_tensor * cov            = ggml_mul_mat(ctx, centered_data, centered_data); // C = X * X^T
+    cov                                 = ggml_scale(ctx, cov, 1.0/(X->ne[0]-1));
+    ggml_build_forward_expand(gf, cov);
+
+    // Step 3: Create ggml_gallocr for graph computation
+    ggml_gallocr_t allocr = ggml_gallocr_new(ggml_backend_get_default_buffer_type(backend));
+    ggml_gallocr_alloc_graph(allocr, gf);
+
+    // Step 4: Check if CPU and compute the result of the graph
+    if (ggml_backend_is_cpu(backend)) {
+        ggml_backend_cpu_set_n_threads(backend, pca_params.n_threads);
+    }
+    ggml_backend_graph_compute(backend, gf);
+
+    // Step 5: Store covariance matrix in the data pointer
+    struct ggml_tensor * result = ggml_graph_node(gf, ggml_graph_n_nodes(gf)-1);
+    float * result_data = (float*) malloc(ggml_nbytes(result));
+    ggml_backend_tensor_get(result, result_data, 0, ggml_nbytes(result));
+    covariance->data = result_data;
+
+    // Step 6: Free memory
+    ggml_gallocr_free(allocr);
+    ggml_free(ctx);
+}
+
+static void compute_cross_covariance(struct pca_params &pca_params,
```

Comment:
Is this normal that this function is unused?

Reply:
Yes, I included this function based on @jukofyork's comments about cross-covariance in the issue discussion. While it’s not currently being utilized, it could be used in future improvements of the algorithm implementation